### PR TITLE
Adding Prettier code formatter

### PR DESCRIPTION
**Pull request by willvedd**
## ✏️ Changes

Adding Prettier to standardize source code formatting with the intended result is to eliminate discussions around style and formatting. Beyond the actual prettier configuration, I've added [pretty-quick](https://www.npmjs.com/package/pretty-quick) to apply formatting and recommit staged files. When paired with Husky, formatting is automatically applied to every commit. 

The rules themselves are based on my own, loosely-held preferences. IMO, the rules are largely irrelevant, consistency and automation are the primary intentions here.

I also needed to tweak our linting quite a bit in order to prevent conflicts stemming from overlapping rules. 

A follow-up PR will be added for that actual implementation of these new style and lint rules.

### CODE REVIEW

File: .prettierignore [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Ignore artifacts:
+local
+lib
+.circleci
+.github
+.husky
+node_modules
+examples
+test/**/*.json
+test/**/*.yaml
```
Comment (sergiughf) [inline .prettierignore:10]: Missing new line.

File: .prettierrc.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+    "semi": false,
+    "trailingComma": "none",
+    "singleQuote": false,
+    "printWidth": 80,
+    "quoteProps": "consistent"
+}
```
Comment (sergiughf) [inline .prettierrc.json:7]: Missing new line.

**Review by sergiughf — APPROVED** [review]
Great addition! 👍🏻 

Apologies for the annoying 2 comments. Feel free to also do that in another PR, as it's not blocking.

File: .husky/pre-commit [hunk 3]
```diff
@@ -3,4 +3,8 @@
 
 npm run lint:fix
 
+npx pretty-quick --staged
+
+npx tsc --noEmit # Ensure that code compiles
```
Comment (willvedd) [inline .husky/pre-commit:8]: One issue I had during testing was that formatting and linting changes would reposition code comments, sometimes resulting in `// @ts-ignore` and `// eslint-disable-foo` being invalidated. Thus, checking that code compiles on precommit can be a good idea.

File: .prettierrc.json (hunk 2 above)
Comment (willvedd) [inline .prettierrc.json:7]: These rules are my own, loosely-held preferences. Happy to modify now, but would hope that we reduce the amount of future modification to reduce noisy git history.

File: package.json [hunk 4]
```diff
@@ -52,11 +53,14 @@
     "cross-env": "^3.1.4",
     "eslint": "^7.28.0",
     "eslint-config-airbnb-base": "^14.2.1",
+    "eslint-config-prettier": "^8.5.0",
```
Comment (willvedd) [inline package.json:56]: This package assists in reducing the level of conflict between eslint and prettier.

File: .prettierrc.json (hunk 2 above)
Comment (cocojoe) [reply .prettierrc.json:7]: It would make sense to align against existing SDKs on the assumption they too have a unified format. @adamjmcgrath ?

Comment (willvedd) [reply .prettierrc.json:7]: That would make sense assuming folks are happy with those rules.

**Comment by willvedd** [comment]
Also, just want to mention that it's expected that this doesn't pass CI because still need to apply the linting and formatting to the entire codebase.

File: .prettierrc.json (hunk 2 above)
Comment (willvedd) [reply .prettierrc.json:7]: So good news about prettier, it applies that rule out-the-box. The follow-up PR will correct this across the entire codebase.

**Review by sergiughf — APPROVED** [review]

**Pull request merged**